src/core/Akka.Persistence/Serialization/PersistenceMessageSerializer.cs

```diff
@@ -236,10 +298,13 @@ private object GetPersistentFSMSnapshot(Type type, byte[] bytes)
                 timeout = TimeSpan.FromMilliseconds(message.TimeoutMillis);
             }
 
+            var payload = GetPayload(message.Data);
+
             // use reflection to create the generic type of PersistentFSM.PersistentFSMSnapshot
-            Type[] types = { typeof(string), type.GenericTypeArguments[0], typeof(TimeSpan?) };
-            object[] arguments = { message.StateIdentifier, GetPayload(message.Data), timeout };
+            Type[] types = { typeof(string), payload.GetType(), typeof(TimeSpan?) };
```

Comment:
Get the type of the payload from the payload type instead of relying on a generic passed into the method